torchgen/dest/register_dispatch_key.py

```diff
@@ -409,6 +416,77 @@ def generate_defn(cpp_sig: CppSignature) -> str:
                 metadata = self.backend_index.get_kernel(f)
                 if metadata is None:
                     return None
+
+                # define the struct for dispatch-less composite kernels.
+                impl_struct = ""
+                if self.backend_index.should_gen_dispatchless_composite(f):
+                    # define the native function as a wrapper static method, redirecting the call to
+                    # the appropriate function (e.g. on CPU, on CUDA, or as a CompositeExplicitAutograd).
+                    def dispatchless_function_defn(
+                        f: NativeFunction, g: Optional[NativeFunctionsGroup]
+                    ) -> str:
+                        dispatch_key = self.backend_index.dispatch_key
+
+                        # the declaration signature is a static method (not a Tensor method).
+                        decl_sig = CppSignatureGroup.from_native_function(
+                            f, method=False, fallback_binding=False
+                        ).signature
+
+                        # decide 2 things:
+                        #     1. where is the redirect function? (i.e. its namespace)
+                        #     2. is it a Tensor method? (some operations are only available as methods)
```

Comment:
I think you should actually be able to clean this up to avoid worrying about the tensor method case.

e.g. `copy_()` is a method-only native function. But it has a generated fast-path function, `at::compositeexplicitautograd::copy_()`, that you should be able to slot into this logic.

Reply:
Hmmm. That's a good point.

However, I think we still need it if, say, a method operation `foo()` has only CPU dispatch but we try to use with a CUDA tensor. If someone registers a kernel for CUDA afterwards, it should still work. Otherwise, if we just eliminate the method case, compilation will fail since `foo()` only has CPU dispatch.

What do you think?

Reply:
Changed my mind (see reply below).